### PR TITLE
Automate straight.el bootstrapping process

### PR DESCRIPTION
**Pull request by declanqian**
Context https://github.com/emacs-ng/emacs-ng/issues/348#issuecomment-890514436

### CODE REVIEW

File: lisp/emacs-lisp/emacs-ng.el [hunk 1]
```diff
@@ -1 +1,112 @@
...
+      (straight-use-recipes
+       '(emacsmirror-mirror :type git :host github
+                            :repo "emacs-straight/emacsmirror-mirror"
+                            :build nil))
+    (straight-use-recipes '(emacsmirror :type git :host github
+                                        :repo "emacsmirror/epkgs"
+                                        :nonrecursive t
+                                        :build nil)))
+
+  (straight-use-package `(straight :type git :host github
```
Comment (declanqian) [inline lisp/emacs-lisp/emacs-ng.el:87]: Should we left this out?

**Comment by declanqian** [comment]
The function names can be improved I think.

**Comment by brotzeit** [comment]
Sorry for the delay, I wasn't able to build it successfully. However, I just ran git clean and now it seems to be working fine.

But I always get dirty submodules. Do you think it makes sense to ignore it with `ignore = dirty` in the submodule config ? Or does this have any side effects we don't want ?

File: lisp/emacs-lisp/emacs-ng.el (hunk 1 above)
Comment (brotzeit) [reply lisp/emacs-lisp/emacs-ng.el:87]: Since we use a submodule for straight, I guess we don't need this.

**Comment by brotzeit** [comment]
Can you please add some notes to https://github.com/emacs-ng/emacs-ng/blob/master/mkdocs.yml#L9.

Just a new section for "use-package and straight" below "Setup". We can improve the docs later, but users should know about it.

**Comment by declanqian** [comment]
> But I always get dirty submodules. Do you think it makes sense to ignore it with ignore = dirty in the submodule config ? Or does this have any side effects we don't want ?

`ignore = dirty` doesn't help in this case. I've added a make `submodule` command to update
submodules. Also, submodules are updated after `make clean` and `./autogen.sh`.
You can use `magit-submodule` to update a individual submodule.

Added some notes in `./admin/notes/git-submodule`.

> Can you please add some notes to

Added a subheading `Built-in Emacs Lisp packages` under `Features` in `README.md` file

> The function names can be improved I think.

Rename `ng/` to `ng-` according to [Emacs Lisp Coding Conventions](https://www.gnu.org/software/emacs/manual/html_node/elisp/Coding-Conventions.html)
I've used `ng` as the short word to distinguish `emacs-ng` from other Lisp programs

**Comment by brotzeit** [comment]
@declanqian sorry again for the delay, it's too hot for me =)

I still had some issues and then I realized that straight doesn't respect a different value for `user-emacs-directory`. There seems to be code in install.el that's necessary.

I think we have to tell the user to put an additional line into his config before installing packages. Or do you see a possibility how we can avoid it ?

**Comment by declanqian** [comment]
Oh, you must do the additional straight lines in the early init file, as the straight bootstrap is done before
loading the regular init file. Just like the way `package-enable-at-startup` works.
@brotzeit 
I have these in my early-init.el file, and it works fine.
```
(setq package-enable-at-startup nil)
(setq straight-repository-branch "develop"
      straight-base-dir (concat user-emacs-directory "var/")
      straight-enable-use-package-integration t
      straight-use-package-by-default t
      straight-build-dir (format
                          "build-%s__%s"
                          emacs-version
                          (number-to-string
                           (time-to-seconds emacs-build-time)))
      straight-cache-autoloads nil
      straight-check-for-modifications nil
      straight-enable-package-integration nil
      straight-vc-git-default-clone-depth 1
      autoload-compute-prefixes nil)
```

**Comment by declanqian** [comment]
Nah, after a second thought. Straight.el only use `user-emacs-directory` as
`straight-base-dir`. If you change `user-emacs-directory` from the `early-init.el`,
`straight.el` should respect you settings.

You also can set `ng-straight-bootstrap-at-startup` to `nil` from early init file, then
manully invoke `ng-bootstrap-straight` afterwards.

`straight.el`'s `bootstrap.el` and `install.el` files have no use here. Probably need a
way to remove them from the final distribution.

**Comment by brotzeit** [comment]
> straight.el's bootstrap.el and install.el files have no use here. Probably need a
way to remove them from the final distribution.

I think we can ignore it. And we can add documentation on how to change `emacs-user-directory` later.

But I still got an issue with loading themes:

```
Debugger entered--Lisp error: (error "Invalid face attribute name" :files)
  internal-set-lisp-face-attribute(default :files ("*.el" ("recipes" "recipes/el-get.rcp") "methods" "el-get-pkg.el") #<frame  *Minibuf-1* - GNU Emacs at brotzeit 0x564c01bd67f0>)
  set-face-attribute(default #<frame  *Minibuf-1* - GNU Emacs at brotzeit 0x564c01bd67f0> :underline nil :overline nil :strike-through nil :box nil :inverse-video nil :stipple nil :inherit nil :extend nil :files ("*.el" ("recipes" "recipes/el-get.rcp") "methods" "el-get-pkg.el") :flavor melpa)
  apply(set-face-attribute default #<frame  *Minibuf-1* - GNU Emacs at brotzeit 0x564c01bd67f0> (:underline nil :overline nil :strike-through nil :box nil :inverse-video nil :stipple nil :inherit nil :extend nil :files ("*.el" ("recipes" "recipes/el-get.rcp") "methods" "el-get-pkg.el") :flavor melpa))
  face-spec-reset-face(default #<frame  *Minibuf-1* - GNU Emacs at brotzeit 0x564c01bd67f0>)
  face-spec-recalc(default #<frame  *Minibuf-1* - GNU Emacs at brotzeit 0x564c01bd67f0>)
  custom-theme-recalc-face(default)
  enable-theme(wombat)
  load-theme(wombat nil nil)
```

When I remove

```
  (straight-use-recipes '(el-get :type git :host github
                                 :repo "dimitri/el-get"
                                 :build nil))
```

everything works fine.

**Comment by brotzeit** [comment]
@declanqian I've tested it again and everything seems to be working. And changing the emacs directory also works with the early init file. I'm actually not even sure I've heard of this before.

**Comment by declanqian** [comment]
Yeah, it happens if we load `emacs-lisp/emacs-ng` from very early stage `loadup`.
It corrupts the elisp `default` symbol somehow. So I've removed that.

Glad things work out for you. I've done changing this branch BTW. @brotzeit

**Review by brotzeit — APPROVED** [review]

**Comment by brotzeit** [comment]
Forgot to approve...

**Comment by brotzeit** [comment]
Maybe rebase on master so tests go green ?

**Review by harryfei — CHANGES_REQUESTED** [review]
If conflict with doom-emacs
```
Warning (straight): Two different recipes given for "straight" (:type cannot be both built-in and git) Disable showing Disable logging
```

**Comment by brotzeit** [comment]
> It conflicts with doom-emacs

I guess we have to fix this in doom-emacs right ?

**Comment by declanqian** [comment]
The warning seems can be emliated with `straight-override-recipe`


``` diff
diff --git a/lisp/emacs-lisp/emacs-ng.el b/lisp/emacs-lisp/emacs-ng.el
index 8d0d8e550e..a2e6f0cb33 100644
--- a/lisp/emacs-lisp/emacs-ng.el
+++ b/lisp/emacs-lisp/emacs-ng.el
@@ -84,8 +84,9 @@ ng-bootstrap-straight
                                         :nonrecursive t
                                         :build nil)))

-  (straight-use-package `(straight :type built-in))
-  (straight-use-package `(use-package :type built-in))
+  ;; Prefer newly included straight and use-package
+  (straight-override-recipe `(straight :type built-in))
+  (straight-override-recipe `(use-package :type built-in))

   (if (straight--modifications 'check-on-save)
       (straight-live-modifications-mode +1)
```


But to be clear, doom is not using the built-in straight.el.

**Comment by brotzeit** [comment]
But we should then probably add a note ?

**Comment by declanqian** [comment]
I think [notes](https://github.com/emacs-ng/emacs-ng/blob/002f7deba9e1494dfbcd1ed7c819e13f0c341550/README.md#usage) here are covered.
For doom distribution to take advantage of the newly builtin straight and use-package,
I can try sent a PR, not sure whether they will accept that.

**Comment by harryfei** [comment]
doom-emacs user could choose to disable `ng-straight-bootstrap-at-startup`. Am I right?

**Review by harryfei — APPROVED** [review]

**Comment by declanqian** [comment]
> doom-emacs user could choose to disable `ng-straight-bootstrap-at-startup`. Am I right?
> Can we add this in our doc too?

There are some occurs in the README.md.
Maybe make this an opt-in by setting `ng-straight-bootstrap-at-startup` to `nil`?

**Comment by declanqian** [comment]
Let's make this optional.

```diff
diff --git a/README.md b/README.md
index d2d7a593fc..646fb86382 100644
--- a/README.md
+++ b/README.md
@@ -125,13 +125,12 @@ We are kind of using `straight.el` as an alternative of `package.el`, providing
 
 ##### Usage
 Emacs NG automatically bootstraps `straight.el` at
-startup if `ng-straight-bootstrap-at-startup` is set to `t`, which is the default.
-You must configure `straight.el` in the early init file,
-as the variable `ng-straight-bootstrap-at-startup` is read before
-loading the regular init file.
+startup if `ng-straight-bootstrap-at-startup` is set to `t`, the default value is `nil`
 
-There are some variables you may be interested in (some of them must be
-set **before** the bootstrap process, if they might affect how
+You must configure the built-in `straight.el` in the early init file,
+as the variable `ng-straight-bootstrap-at-startup` is read before
+loading the regular init file. There are some variables you may be interested in
+(some of them must be set **before** the bootstrap process, if they might affect how
 `straight.el` itself is loaded). You can find the details from [this section](https://github.com/raxod502/straight.el/tree/develop#getting-started)
 of the `straight.el`'s documentation.
 
diff --git a/lisp/emacs-lisp/emacs-ng.el b/lisp/emacs-lisp/emacs-ng.el
index a2e6f0cb33..ed4289c028 100644
--- a/lisp/emacs-lisp/emacs-ng.el
+++ b/lisp/emacs-lisp/emacs-ng.el
@@ -21,7 +21,7 @@ straight-repository-branch
 ;;; Customization options
 
 ;;;###autoload
-(defcustom ng-straight-bootstrap-at-startup t
+(defcustom ng-straight-bootstrap-at-startup nil
   "Whether to bootstrap straight.el when Emacs starts.
 If non-nil, straight.el is bootstrapped before reading the init
 file (but after reading the early init file).  This means that if
```

**Comment by harryfei** [comment]
> > doom-emacs user could choose to disable `ng-straight-bootstrap-at-startup`. Am I right?
> > Can we add this in our doc too?
> 
> There are some occurs in the README.md.
> Maybe make this an opt-in by setting `ng-straight-bootstrap-at-startup` to `nil`?

Frankly, I mean add some specific words for `doom-emacs` issue. If they are facing with this warning, they could find solution in our doc quickly.

**Comment by declanqian** [comment]
How do you configure `early-init` in doom? `ng-straight-bootstrap-at-startup`
has to be set from `early-init.el` stage.
I don't have too much knowledge regarding doom. @harryfei

**Comment by harryfei** [comment]
> How do you configure `early-init` in doom? `ng-straight-bootstrap-at-startup`
> has to be set from `early-init.el` stage.
> I don't have too much knowledge regarding doom. @harryfei

I'm not confirming any solution above in my envrionment. If we can't set `ng-straight-bootstrap-at-startup` early, keeping it optional is good choice.

**Comment by brotzeit** [comment]
> If we can't set ng-straight-bootstrap-at-startup early, keeping it optional is good choice.

I think I agree on this. Maybe we can make this a default later.

**Pull request merged**